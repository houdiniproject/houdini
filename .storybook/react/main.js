// License: LGPL-3.0-or-later
// Based on https://github.com/rails/webpacker/issues/1004#issuecomment-628377930
process.env.NODE_ENV = "development";
const railsWebpackEnv = require("../../config/webpack/environment");

module.exports = {
  stories: [
    "../../app/packs/**/*!(--html).stories.mdx",
    "../../app/packs/**/*!(--html).stories.[tj]s?(x)"
  ],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    'storybook-addon-intl'
  ],
  webpackFinal: (config) => {
    const result = {
    // do mutation to the config
    ...config,
    resolve: {
      ...config.resolve,
      ...railsWebpackEnv.config.resolve,
      modules: railsWebpackEnv.resolvedModules.map((i) => i.value),
    },
    module: {
      ...config.module,
      rules: railsWebpackEnv.loaders
        .filter((i) => !["nodeModules", //not sure why this is here
        "moduleCss" // this addresses issues with our webpack config for css not matching what storybook wants
      ].includes(i.key) )
        .map((i) => i.value),
    },
    plugins: [
      ...config.plugins,
      ...railsWebpackEnv.plugins.map((i) => i.value),
    ],
    
  };

  result.module.rules.filter((i) => i.test.test('.ttf')).forEach((i) => {
    i.use=['url-loader']
  })


  return result;
},
};
