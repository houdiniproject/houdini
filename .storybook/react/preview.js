// License: LGPL-3.0-or-later
import intlDecorate from '../../app/packs/components/tests/intl';
const jest = require('jest-mock');
window.jest = jest;

export const decorators = [intlDecorate()]