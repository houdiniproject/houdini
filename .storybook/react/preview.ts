// License: LGPL-3.0-or-later
import {decorators} from './preview_common';

require('../../app/javascript/components/tests/decorators/roboto')

export {decorators};
