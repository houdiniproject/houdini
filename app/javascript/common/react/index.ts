// License: LGPL-3.0-or-later

/**
 * Adds type safety to your default props.
 *
 * @example
 * // your component
 * function CustomComponent(props:{className:string, disabled?:boolean}) : JSX.Element {
 * 	// not show for
 * }
 *
 * CustomComponent.defaultProps = defaultProps(CustomComponent, {
 * 	className: 'another'
 * })
 *
 * @example <caption>Function properly typechecks</caption>
 * // use the same component as previous example
 * CustomComponent.defaultProps = defaultProps(CustomComponent, {
 * 	className: 'another',
 * 	unexpectedProp: 3 // <-- Typecheck will throw a compile error because you didn't expect this.
 * })
 *
 * @param {(props:TProperty) => JSX.Element} _component - the component you want to create default props for
 * @param {Partial<TProperty>} defaultProps - the defaultProps for your component
 * @returns defaultProps
 * */
export function defaultProps<TProperty>(_component:(props:TProperty) => JSX.Element, defaultProps:Partial<TProperty>):Partial<TProperty> {
	return defaultProps;
}