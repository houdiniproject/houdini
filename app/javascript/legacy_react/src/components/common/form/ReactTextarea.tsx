// License: LGPL-3.0-or-later
import * as React from 'react';
import { observer } from 'mobx-react';
import { injectIntl} from 'react-intl';
import {Field} from "../../../../../../../types/mobx-react-form";
import {InputHTMLAttributes, ReactText, TextareaHTMLAttributes} from "react";
import {action, observable} from "mobx";
import {ReactInputProps} from "./react_input_props";
import {castToNullIfUndef} from "../../../lib/utils";

type InputTypes = ReactInputProps & TextareaHTMLAttributes<HTMLTextAreaElement>


class ReactTextarea extends React.Component<InputTypes, {}> {

  constructor(props:InputTypes){
    super(props)
  }

  @observable
  field:Field


  @action.bound
  componentWillMount(){

    this.field = this.props.field


    this.updateProps()
  }

  componentWillUnmount(){
  }


  componentDidUpdate(prevProps: Readonly<InputTypes>, prevState: Readonly<{}>): void {
    this.updateProps()
  }

  @action.bound
  updateProps() {
    this.field.set('label', castToNullIfUndef(this.props.label))
    this.field.set('placeholder', castToNullIfUndef(this.props.placeholder))
  }

  ///Removes the properties we don't want to put into the input element
  @action.bound
  winnowProps(): InputTypes {
    let ourProps = {...this.props}
    delete ourProps.field
    delete ourProps.value
    return ourProps

  }

  render() {
    return <textarea {...this.winnowProps()} {...this.field.bind()}/>
  }
}

export default observer(ReactTextarea)



