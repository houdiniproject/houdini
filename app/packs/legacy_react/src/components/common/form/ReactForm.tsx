// License: LGPL-3.0-or-later
import * as React from 'react';
import { observer, Provider } from 'mobx-react';
import {Field, Form} from "mobx-react-form";
import {observable, action, toJS} from 'mobx';

export interface ReactFormProps
{
  form:Form
}

///Mostly useless class but, at some point, will replace all our form elements
@observer
export class ReactForm extends React.Component<ReactFormProps, {}> {



  @observable
  form:Form

  @action.bound
  componentWillMount()
  {
    this.form = this.props.form
  }




  componentDidUpdate(){


  }

  componentWillUnmount(){
  }


  render() {

     return <form >
          {this.props.children}
      </form>
  }
}




