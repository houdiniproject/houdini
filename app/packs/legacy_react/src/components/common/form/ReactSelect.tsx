// License: LGPL-3.0-or-later
import * as React from 'react';
import { observer } from 'mobx-react';
import {injectIntl} from 'react-intl';
import {Field} from "../../../../../../../types/mobx-react-form";
import {InputHTMLAttributes} from "react";
import {action, observable} from "mobx";
import {SelectHTMLAttributes} from "react";
import {ReactInputProps} from "./react_input_props";
import {castToNullIfUndef} from "../../../lib/utils";


export interface ReactSelectProps extends ReactInputProps
{
  options?:Array<{id:any, name:string}>
}

type InputTypes = ReactSelectProps & SelectHTMLAttributes<HTMLSelectElement>

class ReactSelect extends React.Component<InputTypes, {}> {

  constructor(props:InputTypes){
    super(props)
  }


  @action.bound
  componentWillMount(){


    this.updateProps()
  }

  componentWillUnmount(){
  }


  componentDidUpdate(prevProps: Readonly<InputTypes >, prevState: Readonly<{}>): void {
    this.updateProps()
  }

  @action.bound
  updateProps() {
    this.props.field.set('label', castToNullIfUndef(this.props.label))
    this.props.field.set('placeholder', castToNullIfUndef(this.props.placeholder))
  }


  ///Removes the properties we don't want to put into the input element
  @action.bound
  winnowProps(): InputTypes {
    let ourProps = {...this.props}
    delete ourProps.field
    delete ourProps.value
    delete ourProps.options
    return ourProps

  }

  render() {

    return <select {...this.winnowProps()} {...this.props.field.bind()}>
      { this.props.options ? this.props.options.map(option =>
        <option key={option.id} value={option.id}>{option.name}</option>
      ) : this.props.children
       }
    </select>

  }
}

export default observer(ReactSelect)



