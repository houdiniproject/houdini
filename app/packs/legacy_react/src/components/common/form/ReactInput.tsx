// License: LGPL-3.0-or-later
import  * as React from 'react';
import { observer, inject, Provider } from 'mobx-react';
import {injectIntl} from 'react-intl';
import {Field} from "mobx-react-form";
import {observable, action, toJS, runInAction} from 'mobx';
import {InputHTMLAttributes} from 'react';
import {ReactInputProps} from "./react_input_props";
import {SelectHTMLAttributes} from "react";
import {ReactSelectProps} from "./ReactSelect";
import {castToNullIfUndef} from "../../../lib/utils";


type InputTypes = ReactInputProps &
  InputHTMLAttributes<HTMLInputElement>

class ReactInput extends React.Component<InputTypes, {}> {

  constructor(props:InputTypes){
    super(props)
  }

  @observable
  field:Field


  @action.bound
  componentWillMount(){

    this.field = this.props.field


    this.updateProps()
  }

  componentWillUnmount(){
  }


  componentDidUpdate(prevProps: Readonly<InputTypes>, prevState: Readonly<{}>): void {
    this.updateProps()
  }

  @action.bound
  updateProps() {
      this.field.set('label', castToNullIfUndef(this.props.label))
      this.field.set('placeholder', castToNullIfUndef(this.props.placeholder))
  }

  ///Removes the properties we don't want to put into the input element
  @action.bound
  winnowProps(): InputTypes {
    let ourProps = {...this.props}
    delete ourProps.field
    delete ourProps.value
    return ourProps

  }

  render() {
      return <input {...this.winnowProps()} {...this.field.bind()}/>
  }
}

export default observer(ReactInput)



