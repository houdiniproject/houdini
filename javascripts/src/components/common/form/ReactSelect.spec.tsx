// License: LGPL-3.0-or-later
import * as React from 'react';
import 'jest';
import {Form} from "mobx-react-form";
import ReactInput from "./ReactInput";
import {ReactForm} from "./ReactForm";
import {action, observable, toJS} from 'mobx';
import ReactTextarea from './ReactTextarea';
import {observer} from 'mobx-react';
import {mount} from 'enzyme';
import ReactSelect from './ReactSelect';


@observer
class TestChange extends React.Component{
  @observable
  remove:boolean
  @observable
  form: Form

  @action.bound
  componentWillMount(){
    this.form = new Form({fields:[{
        name: 'name',
        extra: null}
      ]})
  }



  @action.bound
  onClick(){
    this.remove = true
  }
  render() {
    let reactInput = !this.remove ?  <ReactSelect field={this.form.$('name')} label={'label1'} placeholder={"holder"} options={[{id: null, name:null},
      {id: 'something', name: "Something"},
      {id: 'another_value', name: "aonther value"}
    ]}>

    </ReactSelect> : undefined

    return <ReactForm form={this.form}>

      {reactInput}
      <button onClick={() => this.onClick()}/>
    </ReactForm>
  }
}

describe('ReactSelect', () => {
  let form: Form
  beforeEach(() => {
    form = new Form({
      fields: [
        {
          name: 'name',
          extra: null
        }
      ]
    })
  })

  test('gets added properly', () => {
    let res = mount(<ReactForm form={form}>
      <ReactSelect field={form.$('name')} label={"label"}
                     placeholder={"holder"} value={'snapshot'} aria-required={true}options={[{id: null, name:null},
        {id: 'something', name: "Something"},
        {id: 'another_value', name: "aonther value"}
      ]}/>

    </ReactForm>)


    //Did the attributes settings work as expected back to the objects
    expect(form.$('name').label).toEqual('label')
    expect(form.$('name').placeholder).toEqual('holder')
    expect(form.$('name').value).toEqual('')

    //is the aria attribute passted through to the input
    let input = res.find('select')
    let options = input.find('option')
    expect(options.getElements().length).toBe(3)


    // is the input properly bound?
    input.simulate('change', {target: {value: 'something'}})
    expect(form.$('name').value).toEqual('something')
  })

  test('gets removed properly', () => {

    let res = mount(<TestChange/>)

    // The two casts are needed because Typescript was going blowing up without the 'any' first.
    // Why was it? *shrugs*
    let f = res.find('ReactForm').instance() as any as ReactForm
    expect(f.form.size).toEqual(1)

    res.find('select').simulate('change', {target: {value: 'something'}})

    expect(f.form.$('name').value).toEqual('something')

    res.find('button').simulate('click')
    expect(f.form.size).toEqual(1)

    expect(toJS(res.find('form'))).toMatchSnapshot()

    expect(f.form.$('name').label).toEqual('label1')
    expect(f.form.$('name').placeholder).toEqual('holder')
  })
})