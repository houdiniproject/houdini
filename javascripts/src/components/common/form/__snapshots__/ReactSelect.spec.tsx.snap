// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactSelect gets removed properly 1`] = `
<form>
  <button
    onClick={[Function]}
  />
</form>
`;
